### ISSUE TITLE
performance for multipolygon contains much worse than for geos

### ISSUE THREAD

**Issue by gauteh** [georust/geo#1184]
Hi,

I'm trying to convert some code for roaring-landmask (https://github.com/gauteh/roaring-landmask) from geos to geo (https://github.com/gauteh/roaring-landmask/pull/27). However, checking if a point is contained by a large multipolygon (with all countries from GSHHG) is far slower with geo than with geos. With geos I used `PreparedGeometry`, does something similar exist?

geos:
```
test shapes::tests::benches::test_contains_in_ocean ... bench:       1,586 ns/iter (+/- 45)
test shapes::tests::benches::test_contains_on_land  ... bench:       1,499 ns/iter (+/- 169)
```

geo:
```
test shapes::tests::benches::test_contains_in_ocean ... bench:  10,713,374 ns/iter (+/- 356,454)
test shapes::tests::benches::test_contains_on_land  ... bench:   6,709,172 ns/iter (+/- 255,050)
```

**Comment by urschrei** [comment]
No, but you can decompose the multipolygon into polygons and add them to an rstar spatial index: https://docs.rs/rstar/latest/rstar/struct.RTree.html. The rtree is serializable, re-exported by geo (so it should "just work" for bulk loading the multipolygon Vec), and you can associate extra data with each polygon by using https://docs.rs/rstar/latest/rstar/primitives/struct.GeomWithData.html

**Comment by gauteh** [comment]
Thanks, for the tip. Do you know if that is how geos is doing it?

**Comment by urschrei** [comment]
A geos prepared geometry is an rtree behind the scenes, yes. So things like point in multipolygon query the tree transparently.

**Comment by gauteh** [comment]
I'm trying to add all my Polygons to an RTree, and then check if a point is contained by any of the polygons. But I'm getting this error:

```
   Compiling roaring-landmask v0.7.3 (/home/gauteh/met/roaring-landmask)
error[E0599]: the method `nearest_neighbor` exists for struct `RTree<Polygon>`, but its trait bounds were not satisfied
  --> src/shapes.rs:74:28
   |
74 |         let po = self.geom.nearest_neighbor(&p);
   |                            ^^^^^^^^^^^^^^^^ method cannot be called on `RTree<Polygon>` due to unsatisfied trait bounds
   |
  ::: /home/gauteh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/geo-types-0.7.13/src/geometry/polygon.rs:71:1
   |
71 | pub struct Polygon<T: CoordNum = f64> {
   | ------------------------------------- doesn't satisfy `geo::Polygon: PointDistance`
   |
   = note: the following trait bounds were not satisfied:
           `geo::Polygon: PointDistance`
```

https://github.com/gauteh/roaring-landmask/pull/27/commits/b8e3a21cf7147fe29aa6ed84135c322111ca9485

How can I search through the Polygons and check if the point is contained?

**Comment by gauteh** [comment]
I think `Polygons` should implement `contains_point`, but not `PointDistance`. Only `contains_point` and envelope-stuff is necessary for this check? Maybe `PointDistance` needs to be split.

**Comment by gauteh** [comment]
I wrapped the polygons and implemented contains (https://github.com/gauteh/roaring-landmask/pull/27/commits/60bc592b01c781af591f74fc3229c454e5fbf4c9)

The performance is better, but still roughly 1000x worse than geos :/

```
test shapes::tests::benches::test_contains_in_ocean ... bench:   1,403,390 ns/iter (+/- 175,985)
test shapes::tests::benches::test_contains_on_land  ... bench:   1,395,731 ns/iter (+/- 111,216)
```

**Comment by urschrei** [comment]
Can you post a link to your benchmark? I just want to make sure I'm understanding everything here.

**Comment by gauteh** [comment]
Hi, the benchmark is here: https://github.com/gauteh/roaring-landmask/blob/685ed6f43e944a604545c452273ded4ab901ce42/src/shapes.rs#L201

You run it with: cargo bench shapes --features nightly, requires python at least 3.10. main branch is with geos, the linked PR is with geo. 

Thanks

**Comment by kylebarron** [comment]
FWIW I believe a GEOS prepared geometry indexes _every coordinate pair_ (`Line` in geo terminology). This is vastly different than indexing each polygon, especially for large polygons like country/land boundaries. I think what you want is https://github.com/georust/geo/issues/803. 

I _think_ that geo will automatically build in-effect a prepared geometry before doing the intersection, but there's currently no API to allow the user to persist this. So if you're doing repeated intersections I'd expect bad performance compared to GEOS

**Comment by gauteh** [comment]
Thanks, that seems like something that needs to be done on geo level. It appears to have got some work, but still require some effort?

**Comment by gauteh** [comment]
Would it be enough to persist/cache this intermediate step? Any way to do a quick test if that makes any difference?

**Comment by urschrei** [comment]
You can enable serialise / deserialise support trees by enabling the serde feature on the rstar dependency. I haven't done it myself, but it should be pretty straightforward using something like https://docs.rs/bincode/latest/bincode/

**Comment by gauteh** [comment]
It's not the rstar, but what is mentioned here: https://github.com/georust/geo/issues/1184#issuecomment-2121162800 . Rstar didn't help enough.

**Comment by michaelkirk** [comment]
> Any way to do a quick test if that makes any difference?

As discussed in #803, https://github.com/georust/geo/compare/mkirk/prepared-geom?expand=1 could be a good starting point.

**Comment by gauteh** [comment]
Now testing out prepared geometries through https://github.com/georust/geo/pull/1198, is it better if I create an RTree with prepared polygons, or should I just create one prepared multipolygon? I am testing many points against all the polygons.

**Comment by gauteh** [comment]
Using an RTree of prepared geometries (https://github.com/gauteh/roaring-landmask/pull/29/commits/e41377825fbfb5168c6d30dea9861aa357092985), slightly worse:

```
test shapes::tests::benches::test_contains_in_ocean ... bench:  41,531,670.10 ns/iter (+/- 4,594,429.18)
test shapes::tests::benches::test_contains_on_land  ... bench:  42,088,323.30 ns/iter (+/- 3,958,897.37)
```

**Comment by airbreather** [comment]
I've worked on NTS, the C# port of JTS (whereas GEOS is the C++ port of JTS). Since it looks like part of this discussion focuses **SPECIFICALLY** on the "contains point" case, while it's true that JTS / NTS / GEOS use an R-tree for this, there's some nuance that I don't see being discussed here. *If I've scanned too quickly and missed it, please ignore me...*

For single-point tests specifically, it's NOT a typical 2-dimensional R-tree holding all the line segments. Rather, it uses a 1-dimensional "interval R-tree" of those segments.

Mathematically, calculating "where is `(x, y)` located relative to this (multi/)polygon?", is done by taking a ray that starts at `(x, y)`, following it to infinity in any direction, and counting how many times it crosses a polygon ring segment. Ignoring the "exactly on the boundary" case for this discussion (not because it's unimportant generally, but because it doesn't affect indexing), if the result is even (including zero), then the point is inside. Otherwise, it's outside.

The math doesn't care which direction we choose, but there are four interesting ones: positive / negative, parallel to the x / y axis. [GEOS happens to pick positive X](https://github.com/libgeos/geos/blob/3.13.0/src/algorithm/RayCrossingCounter.cpp#L129-L192), so I'll talk about that specifically, but it can work with any of the four.

Once you commit to choosing choosing that direction, you can index all of the line segments of the (multi/)polygon using a variant of the R-tree that [only looks at their min/max Y values](https://github.com/libgeos/geos/blob/3.13.0/src/algorithm/locate/IndexedPointInAreaLocator.cpp#L73-L82). With that index, a query for point `(x, y)` can cheaply discard any segments that **can't possibly** influence the result because they're either fully above or fully below the query point. What's more, you can very cheaply rule out a lot of the remaining segments as well by [just checking if they're fully on the other side of the ray from your query point](https://github.com/libgeos/geos/blob/3.13.0/src/algorithm/RayCrossingCounter.cpp#L137-L140), leaving a very small subset of the segments, almost all of which actually do need to influence the result in some way almost all of the time.

I hope this helps.

**Comment by PaulWagener** [comment]
The algorithm described by @airbreather seems to be [already implemented](https://github.com/georust/geo/blob/b2dd81111f474c84ee37b5e927dbbf3ae3df8306/geo/src/algorithm/coordinate_position.rs#L388) in geo. Except for the fact that no index is used to iterate through the line segments.

Combining `coord_pos_relative_to_ring()` with an rstar 1-dimensional tree probably gives performance comparable to GEOS. If a maintainer can give some high-level pointers on how best to implement this I'd be happy to create a PR for this.

**Comment by urschrei** [comment]
I don't know whether rstar can be used as described (1D / interval) – this is probably a comprehension failure on my part, but I don't believe it can. Determining this seems like a good first step: there's no fundamental reason we wouldn't adopt a second r-tree library for a single algorithm, but the bar is high (quality of the library, maintenance, effect on compilation speed to name just a few factors).

**Comment by PaulWagener** [comment]
Ah, I assumed that since rstar advertised itself as a "n-dimensional r-tree" it would also work for N = 1, but I'm not familiar enough with that crate to know if it can be used as such.

It looks like something like [locate_in_envelope_intersecting](https://docs.rs/rstar/latest/rstar/struct.RTree.html#method.locate_in_envelope_intersecting) can be used to do an interval query on an RTree

**Comment by michaelkirk** [comment]
For 1-D, could you use a BTree range?https://doc.rust-lang.org/beta/std/collections/btree_map/struct.Range.html

**Comment by urschrei** [comment]
I don't think that will work: btrees (I think) can only order by a single key and that won't work for efficiently finding all intervals that overlap with a query interval (which needs start and end).
I wonder whether we **could** use rstar, if we store the intervals as 0-height rects? I'm not sure whether degenerate geometry is a problem in this case, and I'm sure it's not efficient, but maybe worth experimenting with?

**Comment by urschrei** [comment]
@airbreather Thank you for the detailed suggestion – it's been extremely valuable and I've incorporated the ideas into #1401 wholesale. Feedback on the mechanics and / or some sort of evil corner case test would be very welcome.
@PaulWagener apologies for pouring cold water on the rstar idea – it just hadn't occurred to me that we could use degenerate geometries to sort-of imitate an interval tree like this. It _seems_ to work in terms of the query logic, now all we need is a benchmark.

**Comment by PaulWagener** [comment]
I've [benchmarked](https://github.com/PaulWagener/geo/commit/fb6ec4b8cc52d1158b7f276209164d3c0f7841ea) #1401 and got promising results: 

````
with coordinate_position(): 5.8180 ms
with IndexedMultipolygon: 91.290 µs
with GEOS:  37.502 µs
````

However there is still a bug in there as the IndexedMultipolygon gives a different result on my test dataset than GEOS and the original coordinate_position().

I can increase the performance even further by [using](https://github.com/PaulWagener/geo/commit/6bd9eee13732a1f6f167b823e7d5fa32558a09b1) a true 1D rstar tree:
````
IndexedMultipolygon with 1D rstar: 67.922 µs
````
Though the vanilla crate does not allow 1D. I used [this PR](https://github.com/georust/rstar/pull/169/files) for that benchmark.

Going down to the performance level of GEOS probably requires a more lightweight interval tree.

**Comment by urschrei** [comment]
> However there is still a bug in there as the IndexedMultipolygon gives a different result on my test dataset than GEOS and the original coordinate_position()

Oh that's not good. Can you post a simple repro I can try to incorporate into my testing? Definitely possible that the ray-intersection logic needs to revised.

**Comment by urschrei** [comment]
I also see from the linked comment that https://github.com/adamreichold/sif-itree exists, but I'm struggling to figure out how to use its API with `geo` coords.

**Comment by PaulWagener** [comment]
> Oh that's not good. Can you post a simple repro I can try to incorporate into my testing? Definitely possible that the ray-intersection logic needs to revised.

You can uncomment [this line](https://github.com/PaulWagener/geo/blob/fb6ec4b8cc52d1158b7f276209164d3c0f7841ea/geo/benches/fast_coordinate_position.rs#L42) in the benchmark and use it as a test

**Comment by urschrei** [comment]
@PaulWagener that test now passes as a result of a better (robust) winding number impl. Can you re-run your benchmark with the changes? I suspect we'll be a bit slower due to the robust checks, but as long as we're same-order-of magnitude…

**Comment by urschrei** [comment]
(I get 177 µs vs 7 ms, so 40x)

**Comment by PaulWagener** [comment]
On my system re-running the benchmark I get 87µs vs 5.8ms for a speedup of 66X. Pretty good!

I also [tested](https://github.com/PaulWagener/geo/commit/cf727c18ffff13c10d7bc8179b54be8e42ac3b41) an implementation using the sif_itree crate you mentioned which triples the performance once again to 31 µs. This beats even GEOS on my system which takes 38 µs.

The major downside with sif_itree is that it requires an `Ord` trait bound which f64 does not implement, for the benchmark I used the OrderedFloat workaround.

A merge-worthy implementation should probably include a custom IntervalTree inside of the geo crate. I'll leave it up to the crate maintainers wether the extra performance boost is worth the extra maintenance and compilation speed slowdown.

**Comment by urschrei** [comment]
Hmm sif_itree doesn't add any dependencies (it's only serde and num-traits, which we already require) and we've previously implemented an `Ord`-able `GeoNum` `Coord` where it was needed (https://github.com/georust/geo/blob/main/geo/src/algorithm/old_sweep/point.rs#L7-L62) so we would probably do something similar instead of pulling in `ordered_float`. That might actually simplify matters because we could then swap out the `Contains` impl for MultiPolygon(Point) entirely since `Contains` is for `Geonum`.

**Comment by urschrei** [comment]
(We all pretend to respect the fact that GeoNum and GeoFloat aren't `Ord`, until we need them to be for perf reasons – then we adopt the "well if the data can't actually be ordered we have bigger problems anyway." position)

**Comment by urschrei** [comment]
OK, latest version using sif_itree and `SweepPoint` as a wrapper gets me down to 84 µs (83x speedup over CoordPosition).

**Comment by urschrei** [comment]
Confused about the remaining perf gap compared to https://github.com/PaulWagener/geo/commit/cf727c18ffff13c10d7bc8179b54be8e42ac3b41. Possibilities:
1. Generics (but why?)
2. Conversion to `SweepPoint` in `contains_point` (seemed like the most likely culprit, but I tried pre-computing it and storing it in the struct, and it made no difference)

I also tried removing the `x_max` early rejection code: no effect.

**Comment by PaulWagener** [comment]
I doubt that generics are the cause, but I have some other ideas that could explain the discrepancy:

* The SweepPoint's Ord implementation does a total_cmp() on the X coordinate first. This does nothing because it is always 0, but it is on a hot code path.
* The SweepPoint has to store an extra 0.0 in memory as the X coordinate. Datastructures are generally more efficient when they can fit in less space.

Maybe try implementing something custom like a SweepCoord that only stores a single coordinate? That would solve both problems.

**Comment by urschrei** [comment]
> * The SweepPoint's Ord implementation does a total_cmp() on the X coordinate first. This does nothing because it is always 0, but it is on a hot code path.

This alone gets us to 140x. **Great** suggestion.

**Comment by PaulWagener** [comment]
I've benched your latest implementation on my machine and got a 190x speedup. Looking forward to using this in my project! 

I'll keep my eye on #1401, if I can help let me know

**Comment by gauteh** [comment]
Amazing! Looking forward to testing this when it is ready!
